srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1246 @@
+}
+
+/// Asset creation options.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub struct AssetOptions<Balance: HasCompact, AccountId> {
+	/// Initial issuance of this asset. All deposit to the creater of the asset.
+	#[codec(compact)]
+	pub initial_issuance: Balance,
+	/// Which accounts are allowed to possess this asset.
+	pub permissions: PermissionLatest<AccountId>,
+}
+
+/// Owner of an asset.
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub enum Owner<AccountId> {
+	/// No owner.
+	None,
+	/// Owned by an AccountId
+	Address(AccountId),
+}
+
+impl<AccountId> Default for Owner<AccountId> {
+	fn default() -> Self {
+		Owner::None
+	}
+}
+
+/// Asset permissions
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub struct PermissionsV1<AccountId> {
+	/// Who have permission to update asset permission
+	pub update: Owner<AccountId>,
+	/// Who have permission to mint new asset
+	pub mint: Owner<AccountId>,
+	/// Who have permission to burn asset
+	pub burn: Owner<AccountId>,
+}
+
+/// Versioned asset permission
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+pub enum PermissionVersions<AccountId> {
+	V1(PermissionsV1<AccountId>),
+}
+
+/// Asset permission types
+pub enum PermissionType {
+	/// Permission to update asset permission
+	Burn,
+	/// Permission to mint new asset
+	Mint,
+	/// Permission to burn asset
+	Update,
+}
+
+/// Alias to latest asset permissions
+pub type PermissionLatest<AccountId> = PermissionsV1<AccountId>;
+
+impl<AccountId> Default for PermissionVersions<AccountId> {
+	fn default() -> Self {
+		PermissionVersions::V1(Default::default())
+	}
+}
+
+impl<AccountId> Default for PermissionsV1<AccountId> {
+	fn default() -> Self {
+		PermissionsV1 {
+			update: Owner::None,
+			mint: Owner::None,
+			burn: Owner::None,
+		}
+	}
+}
+
+impl<AccountId> Into<PermissionLatest<AccountId>> for PermissionVersions<AccountId> {
+	fn into(self) -> PermissionLatest<AccountId> {
+		match self {
+			PermissionVersions::V1(v1) => v1,
+		}
+	}
+}
+
+/// Converts the latest permission to other version.
+impl<AccountId> Into<PermissionVersions<AccountId>> for PermissionLatest<AccountId> {
+	fn into(self) -> PermissionVersions<AccountId> {
+		PermissionVersions::V1(self)
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+
+		/// Create a new kind of asset.
+		fn create(origin, options: AssetOptions<T::Balance, T::AccountId>) -> Result {
+			let origin = ensure_signed(origin)?;
+			let id = Self::next_asset_id();
+
+			let permissions: PermissionVersions<T::AccountId> = options.permissions.clone().into();
+
+			// The last available id serves as the overflow mark and won't be used.
+			let next_id = id.checked_add(&One::one()).ok_or_else(||"No new assets id available.")?;
```

Comment:
```suggestion
			let next_id = id.checked_add(&One::one()).ok_or_else(|| "No new assets id available.")?;
```

I would also like to have a better error message here, telling the user that the id would overflow.
@joepetrowski do you have any idea? :)